### ISSUE TITLE
Intermittent "refresh token is invalid" errors

### ISSUE THREAD

**Issue by RPM1984** [IdentityServer/IdentityServer3#2881]
I've been running IdSrv in production for a while (custom flow with password + social grants), works great.

However, sometimes i'm seeing these "Refresh token is invalid" errors in the logs. I've managed to _sometimes_ replicate this on local, but the error doesn't give much insight into _why_ it's invalid:

```
16 May 2016 09:26:29.498  05/15/2016 23:26:28 [Warning] Refresh token is invalid {
  "ClientId": "xxx,
  "ClientName": "xxx,
  "GrantType": "refresh_token",
  "RefreshToken": "cc45dfa6100b6ed78678bb943ac5bb2f",
  "Raw": {
    "grant_type": "refresh_token",
    "refresh_token": "cc45dfa6100b6ed78678bb943ac5bb2f"
  }
}
```

My access token expiry is 1 hour, refresh token sliding absolute exp 7 days.

I'm storing my access/refresh tokens in cookies and refreshing the tokens serverside using `_tokenClient.RequestRefreshTokenAsync`, which is the one that errors with "invalid_grant".

Any ideas on what i can do to debug here?

**Comment by RPM1984** [comment]
@brockallen any ideas on what i can do to debug this?

**Comment by brockallen** [comment]
If your refresh tokens are one-time use and you are refreshing them in parallel only one will "win", so to speak. Perhaps that's it.

**Comment by brockallen** [comment]
All set with this? Can we close?

**Comment by RPM1984** [comment]
I've never solved, as i don't know how to debug as per my previous comment.

**Comment by brockallen** [comment]
Did you ever get any info from the logs, or are they too vague?

**Comment by RPM1984** [comment]
Nope, the only info from the logs is above. Surely there's a way to get more info other than 'refresh token is invalid'? _Why_ is it invalid?

**Comment by brockallen** [comment]
Ok, I can look thru the logging to see if/what we're not omitting and I'll get back to you.

**Comment by leastprivilege** [comment]
Looking at the source code - this log message gets emitted when the refresh token cannot be found in the store. That's where I would investigate

**Comment by johnkors** [comment]
I've experienced the same "issue" some time ago, and it was indeed what @brockallen is mentioning.

> If your refresh tokens are one-time use and you are refreshing them in parallel only one will "win", so to speak.

**Comment by brockallen** [comment]
Do we think we have a conclusion to this question/issue? Can we close?

**Comment by RPM1984** [comment]
@brockallen i think so. I haven't solved, but have enough info/ideas to look into further. Close it. thanks

**Comment by cbrownie** [comment]
@RPM1984 I have been encountering this same issue and been pulling my hair out because of it... did you ever find out why it was happening for you?

**Comment by RPM1984** [comment]
@cbrownie from memory, it was a couple of things (I fixed both, and can't remember which one or if both were cause):
1) Don't try and refresh token in parallel. The first one 'won' then the second errored. This was what @brockallen mentioned. *Fix*: use a singleton to refresh the token.
2) I was in a situation where my client stored the token in a cookie (yes, horrible idea), and there was a mismatch between the cookie expiry and the token expiry. *Fix*: check to ensure grant is valid (eg try/catch and look for `invalid_grant` error), if not log the user out (in my case, clear the cookie)

Pretty sure it was issue 1) causing the error most of the time.

Hope that helps.

**Comment by cbrownie** [comment]
@RPM1984 Thanks for the fast response! Unfortunately I don't think I have parallel attempts to get a refresh token for a user. On the client-side only a single refresh token request is sent at a time (when the access token expires), which sometimes fails. The failure rate on our production server is a lot higher than the failure rate on our development servers, or locally. It;s always the same error you were having, "refresh token is invalid".

I'll have to continue debugging...

**Comment by RPM1984** [comment]
@cbrownie damn. OK well then you might have to enable logging, then open a new issue, and hopefully Brock /Dom can help you out. Good luck!

**Comment by olonga** [comment]
Hi @brockallen/ @johnkors / @RPM1984 ,
Could you please care to explain what you mean by this:
"If your refresh tokens are one-time use and you are refreshing them in parallel only one will "win", so to speak. Perhaps that's it."?
I am using one-time refresh tokens and seem to run into this every now and then.  Not sure what actions are you folks talking about that are being executed in parallel ?

**Comment by johnkors** [comment]
The action of refreshing a token. If you're unlucky, there might be 2 simultanous requests for refreshing tokens. If you've configured them to be one-time use (client setting), only one will succeed.

**Comment by olonga** [comment]
Thanks @johnkors . I have an iframe in shared layout page in mvc app which refreshes the token if it is about to expire. How could this cause 2 simultaneous requests to be sent though?

**Comment by johnkors** [comment]
No idea.
